vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,834 @@
+  gen_quadratic_interactions(stride);
+  // We may also want to 0 out weights here? Currently keep all same in stride position
+}
+
+// This function defines the logic update the live configs' budgets, and to swap out
+// new configs when the budget runs out.
+void config_manager::schedule()
+{
+  for (size_t stride = 0; stride < max_live_configs; ++stride)
+  {
+    if (stride == current_champ) { continue; }
+    if (scores.size() <= stride)
+    {
+      // Skip over removed configs
+      bool removed = true;
+      while (!index_queue.empty() && removed)
+      {
+        removed = configs[index_queue.top().second].state == Removed;
+        if (removed) { index_queue.pop(); }
+      }
+      if (index_queue.empty() && !repopulate_index_queue()) { return; }
+      scored_config sc;
+      sc.config_index = index_queue.top().second;
+      configs[sc.config_index].state = Live;
+      index_queue.pop();
+      scores.push_back(sc);
+      gen_quadratic_interactions(stride);
+    }
+    else if (scores[stride].update_count >= configs[scores[stride].config_index].budget)
+    {
+      handle_empty_budget(stride);
+    }
+  }
+}
+
+void config_manager::update_champ()
+{
+  // Update ips and lower bound for live configs
+  for (size_t stride = 0; stride < scores.size(); ++stride)
+  {
+    float ips = scores[stride].current_ips();
+    distributionally_robust::ScoredDual sd = scores[stride].chisq.recompute_duals();
+    float lower_bound = static_cast<float>(sd.first);
+    configs[scores[stride].config_index].ips = ips;
+    configs[scores[stride].config_index].lower_bound = lower_bound;
+  }
+  float champ_ips = configs[scores[current_champ].config_index].ips;
+  float champ_lower_bound = configs[scores[current_champ].config_index].lower_bound;
+  bool champ_change = false;
+
+  // compare lowerbound of any challenger to the ips of the champ, and switch whenever when the LB beats the champ
+  for (size_t config_index = 0; config_index < configs.size(); ++config_index)
+  {
+    if (configs[config_index].state == New || configs[config_index].state == Removed ||
+        scores[current_champ].config_index == config_index)
+    { continue; }
+    // If challenger is better ('better function from Chacha')
+    if (champ_ips < configs[config_index].lower_bound)
+    {
+      champ_change = true;
+      if (configs[config_index].state == Live)
+      {
+        // Update champ with stride of live config
+        for (size_t stride = 0; stride < scores.size(); ++stride)
+        {
+          if (scores[stride].config_index == config_index)
+          {
+            current_champ = stride;
+            break;
+          }
+        }
+        champ_ips = configs[config_index].ips;
+        champ_lower_bound = configs[config_index].lower_bound;
+      }
+      else
+      {
+        configs[scores[current_champ].config_index].budget *= 2;
+        configs[scores[current_champ].config_index].state = Inactive;
```

Comment:
I suspect this is wrong, if the champ gets over-passed by a single config one that shouldn't kick it that champ out, that should just promote the other one to be active. The hard thing here is that now you have to kick out the worst one.

Reply:
althought the champ might get pushed back in on the next go bc it'll have a high ips probably, but haven't checked that logic

Reply:
Yes this whole `else` branch is the case when an inactive config takes over the champ. I could kick out the worst performing live config instead. That (and handling the index queue more efficiently) are both possible but seem unnecessary since this case really shouldn't happen.

Reply:
synced offline: lower priority since this case might be impossible to hit or uncommon

Reply:
solved the issue in latest commit